### PR TITLE
LLEXT: Relocatable fixes

### PR DESCRIPTION
**Pull request by lyakh**
two fixes for relocatable LLEXT objects. This should fix errors in #72383

### CODE REVIEW

**Review by teburd — DISMISSED** [review]

**Review by pillo79 — COMMENTED** [review]
Thanks for the debug, LGTM, just one small refactor request.

File: subsys/llext/llext.c [hunk 1]
```diff
@@ -564,18 +564,60 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext)
...
+				}
+
+				sym_tab->syms[j].addr = addr + sym.st_value -
+					(ldr->hdr.e_type == ET_REL ? 0 :
+					 shdr.sh_addr);
+				LOG_DBG("%s symbol %d name %s addr %p sect %s #%u",
+					stt == STT_FUNC ? "function" : "object", j, name,
+					sym_tab->syms[j].addr,
+					llext_string(ldr, ext, LLEXT_MEM_SHSTRTAB, shdr.sh_name),
+					sect);
```
[the remark this answers is not in the document]
Comment (lyakh) [reply subsys/llext/llext.c:618]: @pillo79 @teburd they aren't duplicated. Both the `sym_tab->syms[j].addr` assignment and the log entry are different. I can unify the logging call by dropping the difference, but as of now they're different.

[the remark this answers is not in the document]
Comment (pillo79) [reply subsys/llext/llext.c:618]: The only difference in the address calculation (arguably the most important line in the whole function!) is just the name of one argument; in the log, there is a section name that could be calculated for the other case. 
I really see no point in having 2 code paths that do the same logical thing in different ways.

**Review by teburd — COMMENTED** [review]
Minor typo in commit s/stanrads/standard

changes look good to me

**Comment by lyakh** [comment]
> Sorry I maybe wasn't clear enough where I saw the typo, the commit message for [2235022](https://github.com/zephyrproject-rtos/zephyr/commit/22350229951bf29eb731dfee534e614e78369cc7) has "llext: fix handling of non-**stanrad** sections in relocatable case"
> 
> github doesn't necessarily let me comment on a commit message, boooo

@teburd arrrgh, you were perfectly clear, it's just that I cannot type

**Review by teburd — DISMISSED** [review]

File: include/zephyr/llext/llext.h [hunk 2]
```diff
@@ -124,6 +124,8 @@ int llext_iterate(int (*fn)(struct llext *ext, void *arg), void *arg);
 struct llext_load_param {
 	/** Should local relocation be performed */
 	bool relocate_local;
+	/** Has the extension beed relocated to its execution address */
```
Comment (teburd) [inline include/zephyr/llext/llext.h:127]: Small typo here, s/beed/been

File: subsys/llext/llext.c [hunk 3]
```diff
@@ -564,16 +565,57 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext)
...
+				base = llext_peek(ldr, shdr.sh_offset);
+				if (!base) {
+					LOG_ERR("cannot handle arbitrary sections without .peek\n");
+					return -EOPNOTSUPP;
+				}
+
+				section_addr = shdr.sh_addr;
+			}
+
+			if (pre_located)
```
Comment (teburd) [inline subsys/llext/llext.c:611]: Brackets needed

File: subsys/llext/llext.c [hunk 4]
```diff
@@ -564,16 +565,57 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext)
...
+				sym_tab->syms[j].addr = (uint8_t *)sym.st_value +
+					(ldr->hdr.e_type == ET_REL ? section_addr : 0);
+			else
```
Comment (teburd) [inline subsys/llext/llext.c:614]: Brackets please

**Review by teburd — COMMENTED** [review]
Minor nits still, but the change looks ok to me otherwise

**Review by pillo79 — COMMENTED** [review]
I love the idea of being able to respect the ELF-provided addresses, but IMO we really need a test case for this new feature in Zephyr. I don't want to have to _guess_ if any of the patches I submit are breaking SOF... (or even worse, hear screaming coming that way :sweat_smile:)

File: subsys/llext/llext.c [hunk 5]
```diff
@@ -564,16 +564,57 @@ static int llext_copy_symbols(struct llext_loader *ldr, struct llext *ext)
...
+					LOG_ERR("cannot handle arbitrary sections without .peek\n");
+					return -EOPNOTSUPP;
+				}
+
+				section_addr = shdr.sh_addr;
+			}
+
+			if (pre_located)
+				sym_tab->syms[j].addr = (uint8_t *)sym.st_value +
+					(ldr->hdr.e_type == ET_REL ? section_addr : 0);
```
[the remark this answers is not in the document]
Comment (lyakh) [reply subsys/llext/llext.c:613]: @pillo79 no, the difference between `ET_REL` (relocatable object) and `ET_DYN` (shared object) remains. In the relocatable case that calculation is correct:
```
[  113.934401] <inf> llext: llext_copy_symbols: type: 1 sym volume_init: 0x1e8c section 0xa068a000
[  113.934410] <inf> llext: llext_copy_symbols: type: 1 sym volume_func_count: 0x24 section 0xa068e000
[  113.934420] <inf> llext: llext_copy_symbols: type: 1 sym volume_reset_state: 0xf50 section 0xa068a000
```
but in the shared case it's this:
```
[  150.092295] <inf> llext: llext_copy_symbols: type: 3 sym volume_init: 0xa068ba5c section 0xa068a000
[  150.092306] <inf> llext: llext_copy_symbols: type: 3 sym volume_func_count: 0xa068d0b0 section 0xa068d000
[  150.092318] <inf> llext: llext_copy_symbols: type: 3 sym volume_reset_state: 0xa068b510 section 0xa068a000
```

**Comment by lyakh** [comment]
> I love the idea of being able to respect the ELF-provided addresses, but IMO we really need a test case for this new feature in Zephyr.

@pillo79 testing is good... We could add a test where an LLEXT is built for a fixed address, but for which one? You'd probably need to modify a linker script to assign a fixed address range for that extension? And that would be testing / twister-specific, whereas now standard scripts, e.g. soc/cdns/dc233c/include/xtensa-dc233c.ld are used... Any ideas?

**Comment by pillo79** [comment]
> @pillo79 testing is good... We could add a test where an LLEXT is built for a fixed address, but for which one? 

You could use the SOF strategy and add a custom re-link step that only fixes an integer variable to `0xdeadbeef`, then just check its final address from the test. WDYT?

File: subsys/llext/llext.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (pillo79) [reply subsys/llext/llext.c:613]: You're right, that is indeed why it's subtracted in the other case!

**Comment by lyakh** [comment]
> > We could add a test where an LLEXT is built for a fixed address, but for which one?
> 
> You could use the SOF strategy and add a custom re-link step that only fixes an integer variable to `0xdeadbeef`, then just check its final address from the test. (EDIT: If partial placement is not feasible, limit the test to the `qemu-xtensa` arch only). WDYT?

@pillo79 not sure I understand. We need to build an LLEXT with some predefined address, similar to how we do for SOF by using `gcc -Wl,-Tdata=0x........` right? And then QEMU must be able to access that address. So you could do something like:
1. in test_llext_simple.c allocate space for `.data` from the llext object like `uint8_t obj_data[4096];`
2. build everything
3. check zephyr.elf and find the address of `obj_data`
4. re-link LLEXT with `-Wl-Tdata=...`
5. load that LLEXT, copy .data to that array, maybe delete the original .data to make sure the correct one is used...

So, not quite trivial unless there's a simpler way

**Review by pillo79 — APPROVED** [review]
LGTM, will merge the test from #73479 after this.

**Review by teburd — APPROVED** [review]

**Pull request merged**